File: desktop-prototypes/collision-lab-python/collision_model.py
from __future__ import annotations

from dataclasses import dataclass


DEFAULT_WIDTH = 480
DEFAULT_HEIGHT = 320


@dataclass
class Ball:
    id: int
    x: float
    y: float
    radius: float
    mass: float
    velocity_x: float
    color: tuple[int, int, int] = (80, 120, 220)
    name: str = "Ball"


class CollisionLabModel:
    def __init__(self, width: int = DEFAULT_WIDTH, height: int = DEFAULT_HEIGHT) -> None:
        self.width = width
        self.height = height
        self.left_wall = 28.0
        self.right_wall = float(width - 28)
        self.track_y = 122.0
        self.playing = False
        self.status = "paused"
        self.balls: list[Ball] = []
        self.reset()

    def reset(self) -> None:
        self.playing = False
        self.status = "paused"
        self.balls = [
            Ball(1, 145.0, self.track_y, 18.0, 1.0, 70.0, (75, 135, 245), "A"),
            Ball(2, 335.0, self.track_y, 24.0, 2.0, -45.0, (236, 109, 75), "B"),
        ]

    def step(self, dt: float) -> None:
        if not self.playing:
            return

        # Clamp large frame gaps so the simple v0 collision code stays stable.
        dt = max(0.0, min(dt, 1.0 / 20.0))
        for ball in self.balls:
            ball.x += ball.velocity_x * dt

        self._resolve_wall_collisions()
        self._resolve_ball_collision()

    def set_mass(self, ball_id: int, mass: float) -> None:
        ball = self.get_ball(ball_id)
        if ball:
            ball.mass = max(1.0, min(5.0, mass))
            ball.radius = 14.0 + ball.mass * 4.0
            self._separate_balls()
            self._keep_balls_inside()

    def set_velocity(self, ball_id: int, velocity_x: float) -> None:
        ball = self.get_ball(ball_id)
        if ball:
            ball.velocity_x = max(-120.0, min(120.0, velocity_x))

    def get_ball(self, ball_id: int) -> Ball | None:
        return next((ball for ball in self.balls if ball.id == ball_id), None)

    def total_momentum(self) -> float:
        return sum(ball.mass * ball.velocity_x for ball in self.balls)

    def total_kinetic_energy(self) -> float:
        return sum(0.5 * ball.mass * ball.velocity_x * ball.velocity_x for ball in self.balls)

    def _resolve_wall_collisions(self) -> None:
        for ball in self.balls:
            if ball.x - ball.radius < self.left_wall:
                ball.x = self.left_wall + ball.radius
                if ball.velocity_x < 0:
                    ball.velocity_x = -ball.velocity_x
                    self.status = "wall bounce"
            elif ball.x + ball.radius > self.right_wall:
                ball.x = self.right_wall - ball.radius
                if ball.velocity_x > 0:
                    ball.velocity_x = -ball.velocity_x
                    self.status = "wall bounce"

    def _resolve_ball_collision(self) -> None:
        if len(self.balls) < 2:
            return

        left, right = sorted(self.balls, key=lambda ball: ball.x)
        distance = right.x - left.x
        min_distance = left.radius + right.radius
        relative_velocity = left.velocity_x - right.velocity_x

        if distance <= min_distance and relative_velocity > 0:
            v1 = left.velocity_x
            v2 = right.velocity_x
            m1 = left.mass
            m2 = right.mass

            left.velocity_x = ((m1 - m2) / (m1 + m2)) * v1 + ((2.0 * m2) / (m1 + m2)) * v2
            right.velocity_x = ((2.0 * m1) / (m1 + m2)) * v1 + ((m2 - m1) / (m1 + m2)) * v2
            self._separate_pair(left, right)
            self.status = "collision"

    def _separate_balls(self) -> None:
        if len(self.balls) < 2:
            return
        left, right = sorted(self.balls, key=lambda ball: ball.x)
        self._separate_pair(left, right)

    def _separate_pair(self, left: Ball, right: Ball) -> None:
        min_distance = left.radius + right.radius
        overlap = min_distance - (right.x - left.x)
        if overlap <= 0:
            return

        left.x -= overlap / 2.0 + 0.1
        right.x += overlap / 2.0 + 0.1
        self._keep_balls_inside()

    def _keep_balls_inside(self) -> None:
        for ball in self.balls:
            ball.x = max(self.left_wall + ball.radius, min(self.right_wall - ball.radius, ball.x))

File: portable-core/collision-lab/CollisionModel.cpp
#include "CollisionModel.h"

CollisionModel::CollisionModel() {
  reset();
}

void CollisionModel::reset() {
  balls[0].id = 1;
  balls[0].x = 145.0f;
  balls[0].y = trackY;
  balls[0].radius = 18.0f;
  balls[0].mass = 1.0f;
  balls[0].velocityX = 70.0f;

  balls[1].id = 2;
  balls[1].x = 335.0f;
  balls[1].y = trackY;
  balls[1].radius = 24.0f;
  balls[1].mass = 2.0f;
  balls[1].velocityX = -45.0f;

  lastStepCollision = false;
  lastStepWallBounce = false;
}

void CollisionModel::step(float dt) {
  lastStepCollision = false;
  lastStepWallBounce = false;

  dt = clamp(dt, 0.0f, 1.0f / 20.0f);
  for (Ball& ball : balls) {
    ball.x += ball.velocityX * dt;
  }

  handleWallBounce();
  handleBallCollision();
}

void CollisionModel::setMass(int ballId, float mass) {
  Ball& ball = mutableBall(ballId);
  ball.mass = clamp(mass, 1.0f, 5.0f);
  ball.radius = 14.0f + ball.mass * 4.0f;
  separateBalls();
  keepBallsInside();
}

void CollisionModel::setVelocityX(int ballId, float velocityX) {
  mutableBall(ballId).velocityX = clamp(velocityX, -120.0f, 120.0f);
}

const Ball& CollisionModel::getBall(int ballId) const {
  return balls[ballId == 2 ? 1 : 0];
}

float CollisionModel::totalMomentum() const {
  return balls[0].mass * balls[0].velocityX + balls[1].mass * balls[1].velocityX;
}

float CollisionModel::totalKineticEnergy() const {
  return 0.5f * balls[0].mass * balls[0].velocityX * balls[0].velocityX +
         0.5f * balls[1].mass * balls[1].velocityX * balls[1].velocityX;
}

bool CollisionModel::wasLastStepCollision() const {
  return lastStepCollision;
}

bool CollisionModel::wasLastStepWallBounce() const {
  return lastStepWallBounce;
}

Ball& CollisionModel::mutableBall(int ballId) {
  return balls[ballId == 2 ? 1 : 0];
}

void CollisionModel::handleWallBounce() {
  for (Ball& ball : balls) {
    if (ball.x - ball.radius < leftWall) {
      ball.x = leftWall + ball.radius;
      if (ball.velocityX < 0.0f) {
        ball.velocityX = -ball.velocityX;
        lastStepWallBounce = true;
      }
    } else if (ball.x + ball.radius > rightWall) {
      ball.x = rightWall - ball.radius;
      if (ball.velocityX > 0.0f) {
        ball.velocityX = -ball.velocityX;
        lastStepWallBounce = true;
      }
    }
  }
}

void CollisionModel::handleBallCollision() {
  Ball* left = &balls[0];
  Ball* right = &balls[1];
  if (left->x > right->x) {
    left = &balls[1];
    right = &balls[0];
  }

  const float distance = right->x - left->x;
  const float minDistance = left->radius + right->radius;
  const float relativeVelocity = left->velocityX - right->velocityX;

  if (distance <= minDistance && relativeVelocity > 0.0f) {
    const float v1 = left->velocityX;
    const float v2 = right->velocityX;
    const float m1 = left->mass;
    const float m2 = right->mass;
    const float totalMass = m1 + m2;

    left->velocityX = ((m1 - m2) / totalMass) * v1 + ((2.0f * m2) / totalMass) * v2;
    right->velocityX = ((2.0f * m1) / totalMass) * v1 + ((m2 - m1) / totalMass) * v2;
    separatePair(*left, *right);
    lastStepCollision = true;
  }
}

void CollisionModel::separateBalls() {
  Ball* left = &balls[0];
  Ball* right = &balls[1];
  if (left->x > right->x) {
    left = &balls[1];
    right = &balls[0];
  }
  separatePair(*left, *right);
}

void CollisionModel::keepBallsInside() {
  for (Ball& ball : balls) {
    ball.x = clamp(ball.x, leftWall + ball.radius, rightWall - ball.radius);
  }
}

void CollisionModel::separatePair(Ball& left, Ball& right) {
  const float minDistance = left.radius + right.radius;
  const float overlap = minDistance - (right.x - left.x);
  if (overlap <= 0.0f) {
    return;
  }

  left.x -= overlap / 2.0f + 0.1f;
  right.x += overlap / 2.0f + 0.1f;
  keepBallsInside();
}

float CollisionModel::clamp(float value, float minValue, float maxValue) {
  if (value < minValue) {
    return minValue;
  }
  if (value > maxValue) {
    return maxValue;
  }
  return value;
}
